beetsplug/fish.py

```diff
@@ -13,329 +14,393 @@
-        extravalues = opts.extravalues  # e.g., Also complete artists names
-        beetcmds = sorted(
-            (commands.default_commands + commands.plugins.commands()),
-            key=attrgetter("name"),
-        )
-        fields = sorted(set(library.Album.all_keys() + library.Item.all_keys()))
-        # Collect commands, their aliases, and their help text
-        cmd_names_help = []
-        for cmd in beetcmds:
-            names = list(cmd.aliases)
-            names.append(cmd.name)
-            for name in names:
-                cmd_names_help.append((name, cmd.help))
-        # Concatenate the string
-        totstring = HEAD + "\n"
-        totstring += get_cmds_list([name[0] for name in cmd_names_help])
-        totstring += "" if nobasicfields else get_standard_fields(fields)
-        totstring += get_extravalues(lib, extravalues) if extravalues else ""
-        totstring += (
-            "\n"
-            + "# ====== {} =====".format("setup basic beet completion")
-            + "\n" * 2
-        )
-        totstring += get_basic_beet_options()
-        totstring += (
-            "\n"
-            + "# ====== {} =====".format(
-                "setup field completion for subcommands"
-            )
-            + "\n"
-        )
-        totstring += get_subcommands(cmd_names_help, nobasicfields, extravalues)
-        # Set up completion for all the command options
-        totstring += get_all_commands(beetcmds)
-
-        with open(completion_file_path, "w") as fish_file:
-            fish_file.write(totstring)
-
-
-def _escape(name):
-    # Escape ? in fish
-    if name == "?":
-        name = "\\" + name
-    return name
-
-
-def get_cmds_list(cmds_names):
-    # Make a list of all Beets core & plugin commands
-    substr = ""
-    substr += "set CMDS " + " ".join(cmds_names) + ("\n" * 2)
-    return substr
-
-
-def get_standard_fields(fields):
-    # Make a list of album/track fields and append with ':'
-    fields = (field + ":" for field in fields)
-    substr = ""
-    substr += "set FIELDS " + " ".join(fields) + ("\n" * 2)
-    return substr
-
-
-def get_extravalues(lib, extravalues):
-    # Make a list of all values from an album/track field.
-    # 'beet ls albumartist: <TAB>' yields completions for ABBA, Beatles, etc.
-    word = ""
-    values_set = get_set_of_values_for_field(lib, extravalues)
-    for fld in extravalues:
-        extraname = fld.upper() + "S"
-        word += (
-            "set  "
-            + extraname
-            + " "
-            + " ".join(sorted(values_set[fld]))
-            + ("\n" * 2)
+    def run(
+        self,
+        lib: library.Library,
+        opts: optparse.Values,
+        args: List[str],
+    ):
+        # Get the user-provided options.
+        include_fields = not getattr(opts, "noFields")
+        extra_comp_fields = cast(List[str], getattr(opts, "extravalues") or [])
+        output = Path(getattr(opts, "output"))
+        assert len(args) == 0
```

Comment:
```suggestion
        assert not args
```

Reply:
I think it may be worth adding some message for the users here

Reply:
Oh, I didn't think this could be triggered!  I figured the argument parser would throw up an error earlier.  I tested it myself, and it does occur.  I'll add a user-friendly error message.